### ISSUE TITLE
Startup latency/overhead of seqtools.prefetch

### ISSUE THREAD

**Issue by nlgranger** [nlgranger/SeqTools#5]
Moving from #4

seqtools.prefetch creates new workers processes every time it is called which can be resource intensive and negate the advantages of multi-processing.
There is room for optimization if prefetching is called inside a loop on similar objects.

Sample use-case submitted by @pluiez :

```python
import itertools
import random

import numpy
import seqtools


def reader(bufsize=100000):
    with open('a_20GB_text_file') as r:
        while True:
            buffer = [x.strip() for x in itertools.islice(r, bufsize)]
            yield buffer
            if len(buffer) < bufsize:
                break


if __name__ == '__main__':
    word2id = {}


    def preprocess(line):
        words = line.split()
        ids = [word2id[word] for word in words]
        return numpy.array(ids)


    for epoch in range(n):  # repeat `n` epochs
        step_in_epoch = 0
        for text_buffer in reader(100000):  # a single epoch
            random.shuffle(text_buffer)  # impose randomness
            processed = seqtools.smap(preprocess, text_buffer)

            # Run preprocessing in background.
            # However, `prefetch` returns a new `PrefetchedSeqeuence` object that wraps
            # a `AsyncSequence` object, which creates new processes every time when it's initialized.
            fetch = seqtools.prefetch(processed, 10000, 2, 'process')

            while fetch:
                batch = fetch[:batch_size]
                fetch = fetch[batch_size:]
                batch = make_batch(batch)
                # Also, when the segmentation fault or some other critical errors happened in main process,
                # such as during a model running forward and backward on GPU, the sub-processes in `prefetch` 
                # won't be notified to exit. But the torch.utils.data.dataloader can handle these signals 
                # correctly.
                loss = model(batch)
                loss.backward()
                step_in_epoch += 1
```

**Comment by nlgranger** [comment]
@pluiez: I'm moving this to a separate issue because it is unrelated to #4

Currently, prefetch doesn't know anything about the underlying input object, wether it is a single smap, or nested ones, or even a python list. I cannot see a way to specify that a new input is "compatible enough" with another one to reuse a worker.
Moreover, the way multiprocessing works (forking) does not facilitate modifications to the input of a worker: you can't easily serialize a function in python, so you can't serialize an smap, so you can't pass it to a worker after it has started.

It is probably best to try to circumvent the latency issue by moving the prefetch function higher in the pipeline so that is called only once. For example, you can wrap your text file into an object that gives direct access to individual lines or chunks and work with that.

**Comment by pluiez** [comment]
Thank you for your suggestion, I will have a try.

**Comment by nlgranger** [comment]
I was confronted with the issue of updating the input to live workers recently and came up with a trick to update the input of the workers. For 'process' workers, you can allocate an instance of [`multiprocessing.sharedctypes.RawArray`](https://docs.python.org/3/library/multiprocessing.html?highlight=rawarray#multiprocessing.sharedctypes.RawArray) ([wrap it inside a memoryview for optimal performance](https://stackoverflow.com/a/37708824/5786475)) and update it between iterations over the prefetch object. The shared memory also gets updated inside the workers and restarting iterations forces the workers to recompute values based on these updated parameters.

You can see the shuffling mechanism in this example: https://gist.github.com/nlgranger/c02a015af63bd1d18a484ebb34e408b4#file-dataloader-py-L111

The limitation is that RawArray has a fixed size, while your example uses variable length sequences. I think you could compute the lines start end end positions and put that in the shared array to specify which lines should be processed. The workers would then read the data directly from the file at the specified locations.

I think there is no elegant way to work around varying inputs and I'd rather keep seqtools simple to use and understand. I will close this issue for now but feel free to submit more use-cases if you have any, I'll have a look at them.
